### PR TITLE
Remove ISchema in UngroupTransform

### PR DESCRIPTION
**Pull request by wschin**
Remove ISchema in UngroupTransform. It's a part of #1501.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 1]
```diff
@@ -179,21 +179,21 @@ public override bool CanShuffle
...
-            var activeInput = _schemaImpl.GetActiveInput(predicate);
+            var activeInput = _ungroupBinding.GetActiveInput(predicate);
             var inputCursors = Source.GetRowCursorSet(col => activeInput[col], n, null);
             return Utils.BuildArray<RowCursor>(inputCursors.Length,
-                x => new Cursor(Host, inputCursors[x], _schemaImpl, predicate));
+                x => new Cursor(Host, inputCursors[x], _ungroupBinding, predicate));
         }
 
-        private sealed class SchemaImpl : ISchema
+        private sealed class UngroupBinding
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >UngroupBinding [](start = 29, length = 14)

might be a good time to introduce a naming convention for these. No opinion on what it should be, as long as it is consistent :)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 2]
```diff
@@ -254,22 +259,47 @@ public SchemaImpl(IExceptionContext ectx, Schema inputSchema, UngroupMode mode,
...
+                    }
+                    else
+                    {
+                        // i-th input column is a pivot column. Let's calculate proper type and metadata for it.
+                        var metadataBuilder = new MetadataBuilder();
+                        metadataBuilder.Add(inputSchema[i].Metadata, metadataName => ShouldPreserveMetadata(metadataName));
+                        // To explain the output type of pivot columns, let's consider a row
+                        //   Age UserID
+                        //   18  {"Amy", "Willy"}
+                        // where "Age" and "UserID" are column names and 20/{"Amy", "Willy"} is "Age"/"UserID" column in this example row.
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >0 [](start = 74, length = 1)

18, or is 20 the column index?

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 3]
```diff
@@ -337,7 +367,7 @@ public void Save(ModelSaveContext ctx)
             /// </summary>
             public bool[] GetActiveInput(Func<int, bool> predicate)
             {
-                var activeInput = Utils.BuildArray(ColumnCount, predicate);
+                var activeInput = Utils.BuildArray(OutputSchema.Count, predicate);
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >OutputSchema [](start = 51, length = 12)

_inputSchema.Count ?

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 4]
```diff
@@ -359,14 +389,19 @@ public PivotColumnInfo GetPivotColumnInfo(int iinfo)
 
             public PivotColumnInfo GetPivotColumnInfoByCol(int col)
             {
-                _ectx.Assert(0 <= col && col < ColumnCount);
+                _ectx.Assert(0 <= col && col < OutputSchema.Count);
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >OutputSchema.Count [](start = 47, length = 18)

same here

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 5]
```diff
@@ -389,56 +424,11 @@ public int GetCommonPivotColumnSize()
                 }
                 return size;
             }
-
-            public int ColumnCount => _inputSchema.Count;
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >public int ColumnCount => _inputSchema.Count; [](start = 12, length = 45)

why not keep it?

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 6]
```diff
@@ -462,27 +452,27 @@ private sealed class Cursor : LinkedRootCursorBase
...
-            public Cursor(IChannelProvider provider, RowCursor input, SchemaImpl schema, Func<int, bool> predicate)
+            public Cursor(IChannelProvider provider, RowCursor input, UngroupBinding schema, Func<int, bool> predicate)
                 : base(provider, input)
             {
-                _schemaImpl = schema;
-                _active = Utils.BuildArray(_schemaImpl.ColumnCount, predicate);
-                _cachedGetters = new Delegate[_schemaImpl.ColumnCount];
-                _colSizes = new int[_schemaImpl.ColumnCount];
+                _ungroupBinding = schema;
+                _active = Utils.BuildArray(_ungroupBinding.OutputSchema.Count, predicate);
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >Count [](start = 72, length = 5)

would this be the cardinality of the input schema, as well?

**Comment by Ivanidzo4ka** [comment]
>     public sealed class UngroupTransform : TransformBase

I don't see any tests in our repo for this.
Any chance you can port some from internal repo?

---
Refers to: src/Microsoft.ML.Transforms/UngroupTransform.cs:47 in c28125b. [](commit_id = c28125b4b24360dd8fbf47f00b5cee48433fd545, deletion_comment = False)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/UngroupTransform.cs]: It's 18. Thanks.

---
In reply to: [246127382](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246127382) [](ancestors = 246127382)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/UngroupTransform.cs]: `OutputSchema.Count` is the same to `_inputSchema.Count` but I will change it to `_inputSchema.Count`.

---
In reply to: [246128093](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246128093) [](ancestors = 246128093)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/UngroupTransform.cs]: Fixed.

---
In reply to: [246128188](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246128188) [](ancestors = 246128188)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs (hunk 6 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/UngroupTransform.cs]: Input and output schemas contain the same amounts of columns but I will change it to `_ungroupBinding.InputColumnCount` for the sake of clarity.

---
In reply to: [246129434](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246129434) [](ancestors = 246129434)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/UngroupTransform.cs]: Just added it back. Please take a look at `InputColumnCount` in the next iteration.

---
In reply to: [246129129](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246129129) [](ancestors = 246129129)

**Comment by wschin** [comment]
>     public sealed class UngroupTransform : TransformBase

It's tested by entry-point graph.

---
In reply to: [452447487](https://github.com/dotnet/machinelearning/pull/2066#issuecomment-452447487) [](ancestors = 452447487)

---
Refers to: src/Microsoft.ML.Transforms/UngroupTransform.cs:47 in c28125b. [](commit_id = c28125b4b24360dd8fbf47f00b5cee48433fd545, deletion_comment = False)

File: src/Microsoft.ML.Transforms/UngroupTransform.cs [hunk 7]
```diff
@@ -232,44 +232,75 @@ public PivotColumnInfo(string name, int index, int size, PrimitiveType itemType)
             }
 
             private readonly Schema _inputSchema;
+            public int InputColumnCount => _inputSchema.Count;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/UngroupTransform.cs]: >public int InputColumnCount => _inputSchema.Count; [](start = 12, length = 50)

can you move it next to OutputSchema? that layout of public property surrounded of private readonly doesn't looks good.

**Pull request merged**